### DISCUSSION TITLE
Suggestions for system-wide backup

### DISCUSSION THREAD

**Discussion by thomaskamalakis** [Budibase/budibase#5749] [Help] [answered]
Hi all

  Is there a recommended way of backing up all Budibase applications and configuration using the terminal or the API? We are using the self-hosted option.

  Our datasources are on the internal Budibase database, so I guess backing up the couchdB will be sufficient? Could one simply copy the ./budi file (but I guess this should require to stop hosting until the copy is complete to avoid sync issues).
  
  We could always backup the entire VM but this would be a waste of space.

  BR

  Thomas

**Comment by shogunpurple** [comment, marked as answer]
Hi @thomaskamalakis,

First off you may find this discussion here interesting. https://github.com/Budibase/budibase/discussions/4837

It contains several backup options of varying difficulty.

You are correct - the best way is to just back up your couchDB volume, which you could re-attach at will. This is a nice intro:
https://www.spherex.dev/backing-up-docker-volumes/

Being able to do this through our CLI might be a nice addition for self hosters - it could automate quite a lot of this!

**Comment by thomaskamalakis** [comment]
Hi @shogunpurple 

Many thanks

I ended up using ```docker cp``` to copy the contents of all shared volumes ```/opt/couchdb/data/```, etc. So my python script used for **the self-hosted version** for copying the files, zipping the contents and uploading the files to a remote (like Google Drive) look like this:
```
import subprocess
import datetime
import os

def execute(l):
    result = subprocess.run(l, stdout = subprocess.PIPE)
    result.stdout = result.stdout.decode('utf8').split('\n')
    print(result)
    return result

COUCH_CONTAINER_NAME = 'docker_couchdb-service_1'
REDIS_CONTAINER_NAME = 'docker_redis-service_1'
MINIO_CONTAINER_NAME = 'docker_minio-service_1'

ROOT_BACKUP_FOLDER = '/budidisk/docker/couchbackup/'
COUCH_DATA_FOLDER1 = COUCH_CONTAINER_NAME + ':/opt/couchdb/data/.'

REDIS_DATA_FOLDER = REDIS_CONTAINER_NAME + ':/data'
MINIO_DATA_FOLDER = MINIO_CONTAINER_NAME + ':/data'

REMOTE_NAME = 'budibaseBackup'
COUCHDB_SUBFOLDER = 'couchdB'
REDIS_SUBFOLDER = 'redis'
MINIO_SUBFOLDER = 'minio'

now = datetime.datetime.now()
rel_name = now.strftime("%Y-%m-%d-%H-%M-%S")
abs_name = os.path.join(ROOT_BACKUP_FOLDER, rel_name)
target_folder_couch = os.path.join(abs_name, COUCHDB_SUBFOLDER)
target_folder_redis = os.path.join(abs_name, REDIS_SUBFOLDER)
target_folder_minio = os.path.join(abs_name, MINIO_SUBFOLDER)

# make subfolder
execute(['mkdir', '-p', abs_name])

# copy couch
print('backing up:', COUCH_DATA_FOLDER1)
execute(['docker', 'cp', COUCH_DATA_FOLDER1, target_folder_couch])

# copy redis
print('backing up:', REDIS_DATA_FOLDER)
execute(['docker', 'cp', REDIS_DATA_FOLDER, target_folder_redis])

# copy minio
print('backing up:', MINIO_DATA_FOLDER)
execute(['docker', 'cp', MINIO_DATA_FOLDER, target_folder_minio])

# zip subfolder
zip_file = os.path.join(ROOT_BACKUP_FOLDER, rel_name + '.zip')
print('Creating zip file:', zip_file)
execute(['zip', '-r', zip_file, abs_name+'/'])

# sync with Google Drive
print('Uploading to Drive')
execute(['rclone', 'copy', zip_file, REMOTE_NAME + ':/'])
```

Restore on another budibase installation seems to work. You just ```docker cp``` the contents of the folders in the .zip folder to the container volumes of the new installation. I tested this from an 1.0.145 installation to an old 1.0.90 localhost installation and at a first glance app data seemed in tact. I will check to see if everything looks fine after updating to a newer version.

A more elaborate version [here](https://github.com/thomaskamalakis/budibasebackup)

There is a suggestion in the CouchdB [documentation](https://docs.couchdb.org/en/3.2.0/maintenance/backups.html) that one must first copy the ```.shards``` folder and then copy the rest of the contents. 

Thanks again

T.
